File: GainLtc.php
<?php
error_reporting(0);
system('clear');

$cookieFile = __DIR__ . "/cookie.txt";
$configFile = __DIR__ . "/config.json";

// ============================================================
// COLOR
// ============================================================
$R = "\033[0;31m";
$G = "\033[0;32m";
$Y = "\033[0;33m";
$C = "\033[0;36m";
$W = "\033[0;37m";
$RESET = "\033[0m";

// ============================================================
// KONFIGURASI JEDA
// ============================================================
$RUN_DURATION = 10800;  // 3 jam dalam detik
$BREAK_DURATION = 7200; // 120 menit = 7200 detik

// ============================================================
// UI FUNCTIONS
// ============================================================

function print_banner() {
    global $C, $Y, $W, $RESET;
    system('clear');
    echo $C . "========================================================\n";
    echo $C . "║" . $RESET . "                    " . $Y . ":: GainLTC AutoClaim ::" . $RESET . "                     " . $C . "║\n";
    echo $C . "========================================================\n";
    echo $C . "║" . $RESET . " " . $C . "▪ Mode   " . $W . "Auto Claim (Tap-Target Captcha)" . str_repeat(" ", 13) . $C . "║\n";
    echo $C . "║" . $RESET . " " . $C . "▪ Website" . $W . " gainltc.com" . str_repeat(" ", 29) . $C . "║\n";
    echo $C . "========================================================\n";
    echo $C . "--------------------------------------------------------\n\n" . $RESET;
}

function print_box($title, $lines) {
    global $C, $Y, $W, $RESET;
    echo $C . "╔══════════════════════════════════════╗\n";
    echo $C . "║" . $RESET . "         " . $Y . $title . $RESET . "            " . $C . "║\n";
    echo $C . "╚══════════════════════════════════════╝\n";
    foreach ($lines as $line) {
        echo $C . "  " . $line . "\n";
    }
    echo $C . "════════════════════════════════════════\n" . $RESET;
}

function timer($seconds, $label = "⏳ Tunggu") {
    global $Y, $W, $RESET;
    for ($i = $seconds; $i >= 0; $i--) {
        $hours = floor($i / 3600);
        $minutes = floor(($i % 3600) / 60);
        $secs = $i % 60;
        echo "\r" . $Y . $label . ": " . $W . sprintf("%02d:%02d:%02d", $hours, $minutes, $secs) . $RESET . " ";
        flush();
        sleep(1);
    }
    echo "\r" . str_repeat(" ", 60) . "\r";
}

// Hapus cookie lama
@unlink($cookieFile);

// ============================================================
// CONFIG EMAIL & PASSWORD
// ============================================================

print_banner();

$config = [];
if (file_exists($configFile)) {
    $config = json_decode(file_get_contents($configFile), true);
}

if (empty($config['email']) || empty($config['password'])) {
    echo $C . "📧 Email    : " . $W;
    $email = trim(fgets(STDIN));
    echo $C . "🔑 Password : " . $W;
    $password = trim(fgets(STDIN));
    
    file_put_contents($configFile, json_encode([
        "email" => $email,
        "password" => $password
    ], JSON_PRETTY_PRINT));
    
    echo $G . "✅ Berhasil disimpan.\n" . $RESET;
    sleep(2);
    system('clear');
} else {
    $email = $config['email'];
    $password = $config['password'];
}

// ============================================================
// FUNCTION LOGIN
// ============================================================

function do_login($email, $password, $cookieFile) {
    global $C, $Y, $G, $R, $W, $RESET;
    
    // GET login page untuk cookie
    $ch = curl_init();
    curl_setopt_array($ch, [
        CURLOPT_URL => 'https://gainltc.com/login',
        CURLOPT_RETURNTRANSFER => true,
        CURLOPT_COOKIEJAR => $cookieFile,
        CURLOPT_COOKIEFILE => $cookieFile,
        CURLOPT_HTTPHEADER => [
            'User-Agent: Mozilla/5.0 (Linux; Android 10; K) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/150.0.0.0 Mobile Safari/537.36',
            'Accept: text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8',
            'cache-control: max-age=0',
            'sec-ch-ua: "Not;A=Brand";v="8", "Chromium";v="150", "Google Chrome";v="150"',
            'sec-ch-ua-mobile: ?1',
            'sec-ch-ua-platform: "Android"',
            'upgrade-insecure-requests: 1',
            'accept-language: id,en-US;q=0.9,en;q=0.8,pt;q=0.7',
        ],
    ]);
    $response = curl_exec($ch);
    @curl_close($ch);

    // GENERATE CAPTCHA
    $ch = curl_init();
    curl_setopt_array($ch, [
        CURLOPT_URL => 'https://gainltc.com/api/captcha/generate',
        CURLOPT_RETURNTRANSFER => true,
        CURLOPT_POST => true,
        CURLOPT_COOKIEJAR => $cookieFile,
        CURLOPT_COOKIEFILE => $cookieFile,
        CURLOPT_HTTPHEADER => [
            'User-Agent: Mozilla/5.0 (Linux; Android 10; K) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/150.0.0.0 Mobile Safari/537.36',
            'Accept: application/json, text/plain, */*',
            'content-length: 0',
            'sec-ch-ua-platform: "Android"',
            'x-csrf-token: 02672ff5c721b426b153454827446609c5dfdaf1090145235a8e9438762da39a:1785481205783:e393c4975d59889146d05762d0a92f0210bb0d3aa5a4861fa8b9605d2f10b237',
            'sec-ch-ua: "Not;A=Brand";v="8", "Chromium";v="150", "Google Chrome";v="150"',
            'sec-ch-ua-mobile: ?1',
            'origin: https://gainltc.com',
            'accept-language: id,en-US;q=0.9,en;q=0.8,pt;q=0.7',
        ],
    ]);

    while (true) {
        $response = curl_exec($ch);
        $json = json_decode($response, true);
        if (($json['challenge']['type'] ?? '') === 'tap-target') {
            $token = $json['token'];
            $target = $json['challenge']['target'];
            $tapsRequired = $json['challenge']['tapsRequired'];
            break;
        }
        sleep(1);
    }
    @curl_close($ch);

    // VERIFY CAPTCHA
    $ch = curl_init();
    curl_setopt_array($ch, [
        CURLOPT_URL => 'https://gainltc.com/api/captcha/verify',
        CURLOPT_RETURNTRANSFER => true,
        CURLOPT_POST => true,
        CURLOPT_POSTFIELDS => json_encode([
            "token" => $token,
            "answer" => array_fill(0, $tapsRequired, $target),
            "type" => "tap-target"
        ]),
        CURLOPT_COOKIEJAR => $cookieFile,
        CURLOPT_COOKIEFILE => $cookieFile,
        CURLOPT_HTTPHEADER => [
            'User-Agent: Mozilla/5.0 (Linux; Android 10; K) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/150.0.0.0 Mobile Safari/537.36',
            'Accept: application/json, text/plain, */*',
            'Content-Type: application/json',
            'sec-ch-ua-platform: "Android"',
            'x-csrf-token: 02672ff5c721b426b153454827446609c5dfdaf1090145235a8e9438762da39a:1785481205783:e393c4975d59889146d05762d0a92f0210bb0d3aa5a4861fa8b9605d2f10b237',
            'sec-ch-ua: "Not;A=Brand";v="8", "Chromium";v="150", "Google Chrome";v="150"',
            'sec-ch-ua-mobile: ?1',
            'origin: https://gainltc.com',
            'accept-language: id,en-US;q=0.9,en;q=0.8,pt;q=0.7',
        ],
    ]);
    $response = curl_exec($ch);
    @curl_close($ch);
    $json = json_decode($response, true);

    if (empty($json['success'])) {
        return false;
    }
    $verifiedToken = $json['verifiedToken'];

    // LOGIN
    $ch = curl_init();
    curl_setopt_array($ch, [
        CURLOPT_URL => 'https://gainltc.com/api/auth/login',
        CURLOPT_RETURNTRANSFER => true,
        CURLOPT_POST => true,
        CURLOPT_POSTFIELDS => json_encode([
            "email" => $email,
            "password" => $password,
            "captchaToken" => $verifiedToken,
            "rememberMe" => false
        ]),
        CURLOPT_COOKIEJAR => $cookieFile,
        CURLOPT_COOKIEFILE => $cookieFile,
        CURLOPT_HTTPHEADER => [
            'User-Agent: Mozilla/5.0 (Linux; Android 10; K) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/150.0.0.0 Mobile Safari/537.36',
            'Accept: application/json, text/plain, */*',
            'Content-Type: application/json',
            'sec-ch-ua-platform: "Android"',
            'x-csrf-token: 0b2e5587276a2e1e85b4c3b7fe02c135f72a5d7a0329bd88fd77229347a4d231:1785481514850:ff7d5fb947ecf0365600275f3c17173b59e586365390879bbcbf3dfec6c06433',
            'sec-ch-ua: "Not;A=Brand";v="8", "Chromium";v="150", "Google Chrome";v="150"',
            'sec-ch-ua-mobile: ?1',
            'origin: https://gainltc.com',
            'accept-language: id,en-US;q=0.9,en;q=0.8,pt;q=0.7',
        ],
    ]);
    $response = curl_exec($ch);
    @curl_close($ch);
    $json = json_decode($response, true);

    if (empty($json['success'])) {
        return false;
    }

    return $json['user'];
}

// ============================================================
// FUNCTION CLAIM
// ============================================================

function do_claim($cookieFile) {
    global $C, $Y, $G, $R, $W, $RESET;
    
    // GENERATE CAPTCHA
    $ch = curl_init();
    curl_setopt_array($ch, [
        CURLOPT_URL => 'https://gainltc.com/api/captcha/generate',
        CURLOPT_RETURNTRANSFER => true,
        CURLOPT_POST => true,
        CURLOPT_COOKIEJAR => $cookieFile,
        CURLOPT_COOKIEFILE => $cookieFile,
        CURLOPT_HTTPHEADER => [
            'User-Agent: Mozilla/5.0 (Linux; Android 10; K) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/150.0.0.0 Mobile Safari/537.36',
            'Accept: application/json, text/plain, */*',
            'content-length: 0',
            'sec-ch-ua-platform: "Android"',
            'x-csrf-token: 02672ff5c721b426b153454827446609c5dfdaf1090145235a8e9438762da39a:1785481205783:e393c4975d59889146d05762d0a92f0210bb0d3aa5a4861fa8b9605d2f10b237',
            'sec-ch-ua: "Not;A=Brand";v="8", "Chromium";v="150", "Google Chrome";v="150"',
            'sec-ch-ua-mobile: ?1',
            'origin: https://gainltc.com',
            'accept-language: id,en-US;q=0.9,en;q=0.8,pt;q=0.7',
        ],
    ]);

    while (true) {
        $response = curl_exec($ch);
        $json = json_decode($response, true);
        if (($json['challenge']['type'] ?? '') === 'tap-target') {
            $token = $json['token'];
            $target = $json['challenge']['target'];
            $tapsRequired = $json['challenge']['tapsRequired'];
            break;
        }
        sleep(1);
    }
    @curl_close($ch);

    // VERIFY CAPTCHA
    $ch = curl_init();
    curl_setopt_array($ch, [
        CURLOPT_URL => 'https://gainltc.com/api/captcha/verify',
        CURLOPT_RETURNTRANSFER => true,
        CURLOPT_POST => true,
        CURLOPT_POSTFIELDS => json_encode([
            "token" => $token,
            "answer" => array_fill(0, $tapsRequired, $target),
            "type" => "tap-target"
        ]),
        CURLOPT_COOKIEJAR => $cookieFile,
        CURLOPT_COOKIEFILE => $cookieFile,
        CURLOPT_HTTPHEADER => [
            'User-Agent: Mozilla/5.0 (Linux; Android 10; K) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/150.0.0.0 Mobile Safari/537.36',
            'Accept: application/json, text/plain, */*',
            'Content-Type: application/json',
            'sec-ch-ua-platform: "Android"',
            'x-csrf-token: 02672ff5c721b426b153454827446609c5dfdaf1090145235a8e9438762da39a:1785481205783:e393c4975d59889146d05762d0a92f0210bb0d3aa5a4861fa8b9605d2f10b237',
            'sec-ch-ua: "Not;A=Brand";v="8", "Chromium";v="150", "Google Chrome";v="150"',
            'sec-ch-ua-mobile: ?1',
            'origin: https://gainltc.com',
            'accept-language: id,en-US;q=0.9,en;q=0.8,pt;q=0.7',
        ],
    ]);
    $response = curl_exec($ch);
    @curl_close($ch);
    $json = json_decode($response, true);

    if (empty($json['success'])) {
        return false;
    }
    $verifiedToken = $json['verifiedToken'];

    // CLAIM
    $ch = curl_init();
    curl_setopt_array($ch, [
        CURLOPT_URL => 'https://gainltc.com/api/faucet/claim',
        CURLOPT_RETURNTRANSFER => true,
        CURLOPT_POST => true,
        CURLOPT_POSTFIELDS => json_encode([
            "captchaToken" => $verifiedToken
        ]),
        CURLOPT_COOKIEJAR => $cookieFile,
        CURLOPT_COOKIEFILE => $cookieFile,
        CURLOPT_HTTPHEADER => [
            'User-Agent: Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/18.5 Safari/605.1.15',
            'Accept: application/json, text/plain, */*',
            'Content-Type: application/json',
            'sec-ch-ua-platform: "macOS"',
            'x-csrf-token: 25bb0ca451421f370a52901df9ac06919f4df9ec4ed2d16c4b37f36649475c07:1785482476166:010d30483b15235b8f229013182d297431a7bd84822a2c496e0caee18cb22ed0',
            'sec-ch-ua: "Not;A=Brand";v="8", "Chromium";v="150"',
            'sec-ch-ua-mobile: ?0',
            'origin: https://gainltc.com',
            'x-requested-with: mark.via.gp',
            'accept-language: id-ID,id;q=0.9,en-US;q=0.8,en;q=0.7',
        ],
    ]);
    $response = curl_exec($ch);
    @curl_close($ch);
    $json = json_decode($response, true);

    return $json;
}

// ============================================================
// MAIN LOOP
// ============================================================

$start_time = time();
$claim_count = 0;
$total_earned = 0;
$login_attempt = 0;

while (true) {
    
    // Cek apakah sudah jalan 3 jam
    if (time() - $start_time >= $RUN_DURATION) {
        echo $Y . "\n════════════════════════════════════════\n";
        echo $Y . "  ⏸️  Jeda 2 jam (sudah jalan 3 jam)\n";
        echo $Y . "════════════════════════════════════════\n" . $RESET;
        timer($BREAK_DURATION, "⏳ Jeda");
        $start_time = time();
        continue;
    }
    
    print_banner();
    
    echo $C . "📧 Email    : " . $W . $email . "\n";
    echo $C . "📊 Claims   : " . $W . $claim_count . "\n";
    echo $C . "💰 Earned   : " . $G . number_format($total_earned, 8) . "\n";
    echo $C . "⏱️  Running  : " . $W . gmdate("H:i:s", time() - $start_time) . " / 3:00:00\n";
    echo $C . "════════════════════════════════════════\n" . $RESET;
    
    // Login
    if ($login_attempt == 0) {
        echo $Y . "🔑 Login...\n" . $RESET;
    } else {
        echo $Y . "🔑 Re-login (attempt $login_attempt)...\n" . $RESET;
    }
    
    $user = do_login($email, $password, $cookieFile);
    if (!$user) {
        $login_attempt++;
        echo $R . "❌ Login gagal, coba ulang dalam 10 detik...\n" . $RESET;
        sleep(10);
        continue;
    }
    $login_attempt = 0;
    
    echo $G . "✅ Login berhasil! Balance: " . $user['balance'] . "\n" . $RESET;
    sleep(2);
    
    // Claim loop
    $claim_attempts = 0;
    $max_attempts = 10;
    
    while ($claim_attempts < $max_attempts) {
        
        // Cek jeda 3 jam
        if (time() - $start_time >= $RUN_DURATION) {
            break 2;
        }
        
        $result = do_claim($cookieFile);
        
        if (isset($result['roll'])) {
            $claim_count++;
            $reward = floatval($result['reward']);
            $total_earned += $reward;
            $user['balance'] = $result['newBalance'];
            
            echo $G . "╔══════════════════════════════════════╗\n";
            echo $G . "║          🎲 CLAIM BERHASIL          ║\n";
            echo $G . "╚══════════════════════════════════════╝\n";
            echo $C . "🎲 Roll         : " . $W . $result['roll'] . "\n";
            echo $C . "🎁 Reward       : " . $G . $result['reward'] . "\n";
            echo $C . "💰 New Balance  : " . $G . $result['newBalance'] . "\n";
            echo $C . "📊 Total Claims : " . $W . $claim_count . "\n";
            echo $C . "════════════════════════════════════════\n" . $RESET;
            
            $claim_attempts = 0;
            
            $nextClaimAt = intval($result['nextClaimAt'] / 1000);
            $remaining = max(0, $nextClaimAt - time());
            timer($remaining, "⏳ Next claim");
            
        } else {
            $claim_attempts++;
            $err_msg = $result['error'] ?? 'Unknown error';
            echo $R . "❌ Claim gagal ($claim_attempts/$max_attempts): $err_msg\n" . $RESET;
            
            if ($claim_attempts >= $max_attempts) {
                echo $Y . "⚠️  Terlalu banyak gagal, re-login...\n" . $RESET;
                break;
            }
            
            sleep(5);
        }
    }
}
?>

File: GainLtc.py
#!/usr/bin/env python3
"""
GainLTC Auto Claim Bot - Full Color & Format
Support: count, pattern, sequence, emoji-slider, tap-target, drag-order, connect-pairs
"""

import requests
import time
import json
import sys
from datetime import datetime, timedelta
from typing import Optional, Dict, Any, List

# ========== WARNA ANSI ==========
RESET = "\033[0m"
BOLD = "\033[1m"
RED = "\033[91m"
GREEN = "\033[92m"
YELLOW = "\033[93m"
BLUE = "\033[94m"
MAGENTA = "\033[95m"
CYAN = "\033[96m"
WHITE = "\033[97m"
GRAY = "\033[90m"

# ========== KONFIGURASI ==========
BASE_URL = "https://gainltc.com"
RUN_HOURS = 8
MAX_CONSECUTIVE_FAILS = 5

# ========== SESSION ==========
session = requests.Session()
session.headers.update({
    "User-Agent": "Mozilla/5.0 (Linux; Android 10; K) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/127.0.0.0 Mobile Safari/537.36",
    "Accept": "application/json, text/plain, */*",
    "Accept-Language": "id-ID,en;q=0.9",
    "Origin": BASE_URL,
    "Referer": f"{BASE_URL}/",
    "Sec-Fetch-Site": "same-origin",
    "Sec-Fetch-Mode": "cors",
    "Sec-Fetch-Dest": "empty",
})

# ========== PRINT FUNCTIONS ==========
def print_header():
    print(f"\n{BLUE}{'='*60}{RESET}")
    print(f"{BOLD}{CYAN}Website : {WHITE}GainLTC.com{RESET}")
    print(f"{BOLD}{CYAN}👨‍💻 ScriptMaker : {WHITE}@JoshuaXSupport{RESET}")
    print(f"{BOLD}{CYAN}📢 TG          : {WHITE}https://t.me/+f3QBLkR5D8k4YzNl{RESET}")
    print(f"{BLUE}{'='*60}{RESET}\n")

def print_info(msg):
    print(f"{BLUE}🚀 [{datetime.now().strftime('%H:%M:%S')}] {BOLD}[INFO]{RESET} {msg}")

def print_ok(msg):
    print(f"{GREEN}✅ [{datetime.now().strftime('%H:%M:%S')}] {BOLD}[ OK ]{RESET} {msg}")

def print_wait(msg):
    print(f"{YELLOW}⏳ [{datetime.now().strftime('%H:%M:%S')}] {BOLD}[WAIT]{RESET} {msg}")

def print_error(msg):
    print(f"{RED}❌ [{datetime.now().strftime('%H:%M:%S')}] {BOLD}[ERR ]{RESET} {msg}")

def print_sep():
    print(f"{GRAY}{'='*60}{RESET}")

# ========== FUNGSI BANTU ==========
def wait_until(target_ts: int):
    now = int(time.time() * 1000)
    if target_ts > now:
        wait_sec = (target_ts - now) / 1000.0
        print_wait(f"Cooldown {wait_sec:.1f} detik")
        time.sleep(wait_sec + 1)

def get_csrf_token() -> Optional[str]:
    try:
        resp = session.get(f"{BASE_URL}/api/csrf-token")
        if resp.status_code == 200:
            data = resp.json()
            if data.get("success"):
                csrf_cookie = session.cookies.get("csrf_token")
                if csrf_cookie:
                    return csrf_cookie.split(":")[0]
                else:
                    print_error("CSRF token tidak ditemukan di cookie")
                    return None
            else:
                print_error(f"Gagal get CSRF: {data}")
                return None
        else:
            print_error(f"HTTP {resp.status_code} saat get CSRF")
            return None
    except Exception as e:
        print_error(f"Error get CSRF: {e}")
        return None

def generate_captcha() -> Optional[Dict[str, Any]]:
    try:
        resp = session.post(f"{BASE_URL}/api/captcha/generate", json={})
        if resp.status_code == 200:
            data = resp.json()
            if data.get("success"):
                return data
            else:
                print_error(f"Generate captcha gagal: {data}")
                return None
        else:
            print_error(f"HTTP {resp.status_code} generate captcha")
            return None
    except Exception as e:
        print_error(f"Error generate captcha: {e}")
        return None

def solve_captcha(challenge: Dict[str, Any]) -> List[str]:
    ctype = challenge.get("type")
    if ctype == "count":
        winner = challenge.get("winner")
        if winner:
            return [winner]
        else:
            print_error("Tidak ada winner pada count captcha")
            return []
    
    elif ctype in ("pattern", "sequence"):
        seq = challenge.get("sequence", [])
        null_idx = next((i for i, v in enumerate(seq) if v is None), -1)
        if null_idx == -1:
            print_error("Tidak ada null pada pattern/sequence")
            return []
        non_null = [v for v in seq if v is not None]
        if not non_null:
            return []
        # Cari siklus terkecil
        cycle_len = 1
        found = False
        for cl in range(1, len(non_null) + 1):
            pattern = non_null[:cl]
            ok = True
            for i in range(cl, len(non_null)):
                if non_null[i] != pattern[i % cl]:
                    ok = False
                    break
            if ok:
                cycle_len = cl
                found = True
                break
        if not found:
            cycle_len = len(non_null)
            pattern = non_null
        else:
            pattern = non_null[:cycle_len]
        missing = pattern[null_idx % cycle_len]
        return [missing]
    
    elif ctype == "emoji-slider":
        active = challenge.get("activeEmoji")
        ghosts = challenge.get("ghosts", [])
        for g in ghosts:
            if g.get("emoji") == active:
                return [str(g.get("position"))]
        print_error("Tidak menemukan activeEmoji di ghosts")
        return []
    
    elif ctype == "tap-target":
        target = challenge.get("target")
        taps = challenge.get("tapsRequired", 1)
        if target:
            return [target] * taps
        else:
            print_error("Tidak ada target pada tap-target")
            return []
    
    elif ctype == "drag-order":
        correct = challenge.get("correctOrder")
        if correct:
            return correct
        else:
            print_error("Tidak ada correctOrder pada drag-order")
            return []
    
    elif ctype == "connect-pairs":
        left = challenge.get("leftOrder")
        if left:
            return left
        else:
            print_error("Tidak ada leftOrder pada connect-pairs")
            return []
    
    else:
        print_error(f"Tipe captcha tidak dikenal: {ctype}")
        print_error(f"Challenge structure: {json.dumps(challenge, indent=2)}")
        return []

def verify_captcha(token: str, answer: List[str], ctype: str) -> Optional[str]:
    payload = {
        "token": token,
        "answer": answer,
        "type": ctype
    }
    try:
        resp = session.post(f"{BASE_URL}/api/captcha/verify", json=payload)
        if resp.status_code == 200:
            data = resp.json()
            if data.get("success"):
                return data.get("verifiedToken")
            else:
                print_error(f"Verify captcha gagal: {data}")
                return None
        else:
            print_error(f"HTTP {resp.status_code} verify captcha")
            return None
    except Exception as e:
        print_error(f"Error verify captcha: {e}")
        return None

def login(email: str, password: str) -> bool:
    print_info("Login...")
    csrf = get_csrf_token()
    if not csrf:
        print_error("Gagal mendapatkan CSRF token")
        return False
    session.headers.update({"X-CSRF-Token": csrf})

    captcha_data = generate_captcha()
    if not captcha_data:
        print_error("Gagal generate captcha")
        return False
    token = captcha_data.get("token")
    challenge = captcha_data.get("challenge")
    if not token or not challenge:
        print_error("Captcha data tidak lengkap")
        return False
    ctype = challenge.get("type", "count")

    answer = solve_captcha(challenge)
    if not answer:
        print_error("Gagal memecahkan captcha")
        return False

    verified = verify_captcha(token, answer, ctype)
    if not verified:
        print_error("Gagal verify captcha")
        return False

    login_payload = {
        "email": email,
        "password": password,
        "captchaToken": verified,
        "rememberMe": True
    }
    try:
        resp = session.post(f"{BASE_URL}/api/auth/login", json=login_payload)
        if resp.status_code == 200:
            data = resp.json()
            if data.get("success"):
                print_ok("Login berhasil")
                return True
            else:
                print_error(f"Login gagal: {data}")
                return False
        else:
            print_error(f"HTTP {resp.status_code} saat login")
            return False
    except Exception as e:
        print_error(f"Error login: {e}")
        return False

def get_faucet_status() -> Optional[Dict[str, Any]]:
    try:
        resp = session.get(f"{BASE_URL}/api/faucet/status")
        if resp.status_code == 200:
            data = resp.json()
            if data.get("canClaim") is not None:
                return data
            else:
                print_error(f"Status faucet tidak valid: {data}")
                return None
        else:
            print_error(f"HTTP {resp.status_code} get faucet status")
            return None
    except Exception as e:
        print_error(f"Error get faucet status: {e}")
        return None

def claim_faucet() -> Optional[Dict[str, Any]]:
    captcha_data = generate_captcha()
    if not captcha_data:
        print_error("Gagal generate captcha untuk claim")
        return None
    token = captcha_data.get("token")
    challenge = captcha_data.get("challenge")
    if not token or not challenge:
        print_error("Captcha data tidak lengkap untuk claim")
        return None
    ctype = challenge.get("type", "count")

    answer = solve_captcha(challenge)
    if not answer:
        print_error("Gagal memecahkan captcha untuk claim")
        return None

    verified = verify_captcha(token, answer, ctype)
    if not verified:
        print_error("Gagal verify captcha untuk claim")
        return None

    payload = {"captchaToken": verified}
    try:
        resp = session.post(f"{BASE_URL}/api/faucet/claim", json=payload)
        if resp.status_code == 200:
            data = resp.json()
            if "roll" in data and "reward" in data:
                return data
            else:
                print_error(f"Claim response tidak lengkap: {data}")
                return None
        else:
            print_error(f"HTTP {resp.status_code} saat claim")
            return None
    except Exception as e:
        print_error(f"Error claim: {e}")
        return None

# ========== MAIN ==========
def main():
    print_header()

    print_info(f"Target  : {RUN_HOURS} Jam")

    email = input(f"{CYAN}📧 Masukkan email GainLTC Anda: {RESET}").strip()
    if not email:
        print_error("Email tidak boleh kosong. Keluar.")
        sys.exit(1)
    password = input(f"{CYAN}🔑 Masukkan password GainLTC Anda: {RESET}").strip()
    if not password:
        print_error("Password tidak boleh kosong. Keluar.")
        sys.exit(1)

    print_sep()

    if not login(email, password):
        print_error("Login gagal, keluar")
        sys.exit(1)

    print_sep()

    start_time = datetime.now()
    end_time = start_time + timedelta(hours=RUN_HOURS)
    claim_count = 0
    consecutive_fails = 0

    while datetime.now() < end_time:
        status = get_faucet_status()
        if not status:
            print_wait("Gagal mendapatkan status, coba lagi 5 detik")
            time.sleep(5)
            continue

        can_claim = status.get("canClaim", False)
        next_claim_at = status.get("nextClaimAt")

        if not can_claim and next_claim_at:
            wait_until(next_claim_at)
            continue
        elif not can_claim:
            print_wait("Tidak bisa claim dan tidak ada nextClaimAt, tunggu 30 detik")
            time.sleep(30)
            continue

        claim_count += 1
        print_info(f"Claim #{claim_count} dimulai")
        result = claim_faucet()
        if result:
            reward = result.get("reward", 0)
            new_balance = result.get("newBalance", 0)
            next_claim = result.get("nextClaimAt")
            consecutive_fails = 0
            print_ok(f"Reward +{reward}⭐️ | 💳 Balance {new_balance}⭐️")
            if next_claim:
                wait_until(next_claim)
            else:
                print_wait("Tidak ada nextClaimAt, tunggu 10 detik")
                time.sleep(10)
        else:
            consecutive_fails += 1
            print_wait(f"Claim gagal ({consecutive_fails}/{MAX_CONSECUTIVE_FAILS})")
            if consecutive_fails >= MAX_CONSECUTIVE_FAILS:
                print_wait("Terlalu banyak gagal, mencoba re-login...")
                if not login(email, password):
                    print_error("Re-login gagal, keluar")
                    break
                consecutive_fails = 0
            else:
                time.sleep(10)

        # Cek session
        check = get_faucet_status()
        if check is None:
            print_wait("Session mungkin expired, login ulang...")
            if not login(email, password):
                print_error("Login ulang gagal, keluar")
                break

        print_sep()

    print_info(f"Bot selesai setelah {RUN_HOURS} jam. Total claim: {claim_count}")

if __name__ == "__main__":
    main()
